build.c

```diff
@@ -290,9 +306,31 @@ jobstart(struct job *j, struct edge *e)
 		description = buildopts.verbose ? NULL : edgevar(e, "description", true);
 		if (!description || description->n == 0)
 			description = j->cmd;
-		formatstatus(status, sizeof(status));
+		if (!dumb)
+			putchar('\r');
+		len = formatstatus(status, sizeof(status));
 		fputs(status, stdout);
-		puts(description->s);
+#if HAVE_TIOCGWINSZ
+		ioctl(1, TIOCGWINSZ, &ws);
+		if (!dumb) {
+			if (len + description->n > ws.ws_col) {
+				switch (ws.ws_col) {
+					case 3: putchar('.');  /* fallthrough */
+					case 2: putchar('.');  /* fallthrough */
+					case 1: putchar('.');  /* fallthrough */
+					case 0: break;
+					default: printf("%.*s...%s",
+							(ws.ws_col - 3) / 2 - len, description->s,
```

Comment:
Does ninja do anything related to multi-byte character handling here, or does it just assume one byte == one column? If ninja doesn't do anything fancy, than I guess that's okay here, too.

Reply:
As far as I can tell, ninja uses C++'s string::size for this, which apparently returns a byte-count, not taking multi-byte characters into account, so I think it's fine.